### PR TITLE
[BugFix] Late recycle if repair tablet from recycle bin

### LINKED ISSUE #6776: FE exists with npe when replaying add replica
### Steps to reproduce the behavior (Required)
1. set catalog_trash_expire_second to 60 in fe.conf
2. drop tables and restart be continually.

### Expected behavior (Required)
```
2022-05-31 19:15:54,512 INFO (stateListener|76) [CatalogRecycleBin.replayEraseTable():338] replay erase table[6057015] finished
2022-05-31 19:15:54,513 ERROR (stateListener|76) [EditLog.loadJournal():834] Operation Type 42
java.lang.NullPointerException: null
        at com.starrocks.server.LocalMetastore.getPartitionIncludeRecycleBin(LocalMetastore.java:2636) ~[starrocks-fe.jar:?]
        at com.starrocks.server.LocalMetastore.unprotectAddReplica(LocalMetastore.java:2507) ~[starrocks-fe.jar:?]
        at com.starrocks.server.LocalMetastore.replayAddReplica(LocalMetastore.java:2542) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayAddReplica(GlobalStateMgr.java:2236) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.EditLog.loadJournal(EditLog.java:387) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:1696) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.transferToMaster(GlobalStateMgr.java:902) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.access$800(GlobalStateMgr.java:259) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$4.runOneCycle(GlobalStateMgr.java:1618) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
```

### StarRocks version (Required)
 - all

### PR DESCRIPTION
**Pull request by padmejin**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #6776

## Problem Summary(Required) ：

Submitting repair tasks only if no expired db/table/partition involved to avoid NPE when tablet is added after its db/table/partition is dropped.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 48 / 58 (82.76%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|26|32|81.25%|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|22|26|84.62%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 48 / 55 (87.27%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|22|26|84.62%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|26|29|89.66%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 1]
```diff
@@ -253,7 +293,7 @@ private synchronized void eraseDatabaseWithSameName(String dbName) {
             Database db = dbInfo.getDb();
             if (db.getFullName().equals(dbName)) {
                 iterator.remove();
-                idToRecycleTime.remove(entry.getKey());
+                removeRecycleMarkers(entry.getKey());
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:296]: Your solution can not cover this case: Drop a table and create a same name table, the dropped table must be erased immediately, even thought it has been marked as erase later.

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 2]
```diff
@@ -77,7 +78,12 @@ public class CatalogRecycleBin extends MasterDaemon implements Writable {
     private com.google.common.collect.Table<Long, String, RecycleTableInfo> nameToTableInfo;
     private Map<Long, RecyclePartitionInfo> idToPartition;
 
-    private Map<Long, Long> idToRecycleTime;
+    // if enable erase later, the real recycle time is RecycleTime + LATE_RECYCLE_INTERVAL_SECONDS
+    // This is only useful on master, when tablet scheduler repair a tablet that is about to expire
+    // We can make the db/table/partition infomation stay logger until the asynchronized agent task finish.
+    protected static int LATE_RECYCLE_INTERVAL_SECONDS = 10;
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:84]: This value can be longer, because the clone may cost a longer time.

Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java]: How about 60?

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 3]
```diff
@@ -217,22 +229,50 @@ public synchronized List<Partition> getPartitions(long tableId) {
...
+        // 1. not in idToRecycleTime, maybe already erased, sorry it's too late!
+        if (!idToRecycleTime.containsKey(id)) {
+            return false;
+        }
+        // 2. will expire after quite a long time, don't worry
+        long latency = currentTimeMs - idToRecycleTime.get(id);
+        if (latency < (Config.catalog_trash_expire_second - LATE_RECYCLE_INTERVAL_SECONDS) * 1000L) {
+            return true;
+        }
+        enableEraseLater.add(id);
```
[the remark this answers is not in the document]
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java]: Done.

File: fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java [hunk 4]
```diff
@@ -477,16 +479,9 @@ private void schedulePendingTablets() {
...
-        for (AgentTask task : batchTask.getAllTasks()) {
-            if (AgentTaskQueue.addTask(task)) {
-                stat.counterCloneTask.incrementAndGet();
-            }
-            LOG.info("add clone task to agent task queue: {}", task);
-        }
-
-        // send task immediately
-        AgentTaskExecutor.submit(batchTask);
+        // check all tasks, only submit the tasks on tablets that won't expired and erased
```
[the remark this answers is not in the document]
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java]: Done.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 40 / 46 (86.96%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|14|17|82.35%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|26|29|89.66%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 40 / 46 (86.96%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|14|17|82.35%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|26|29|89.66%|

**Review by gengjun-git — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 5]
```diff
@@ -77,7 +78,12 @@ public class CatalogRecycleBin extends MasterDaemon implements Writable {
     private com.google.common.collect.Table<Long, String, RecycleTableInfo> nameToTableInfo;
     private Map<Long, RecyclePartitionInfo> idToPartition;
 
-    private Map<Long, Long> idToRecycleTime;
+    // if enable erase later, the real recycle time is RecycleTime + LATE_RECYCLE_INTERVAL_SECONDS
+    // This is only useful on master, when tablet scheduler repair a tablet that is about to expire
+    // We can make the db/table/partition infomation stay longer until the asynchronized agent task finish.
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:83]: spell error. `information`

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 6]
```diff
@@ -217,22 +229,50 @@ public synchronized List<Partition> getPartitions(long tableId) {
...
 
-    private synchronized boolean isExpire(long id, long currentTimeMs) {
+    /**
+     * if we can erase this instance, we should check if anyone enable erase later.
+     * Only used by main loop.
+     */
+    private synchronized boolean canErase(long id, long currentTimeMs) {
         long latency = currentTimeMs - idToRecycleTime.get(id);
+        if (enableEraseLater.contains(id)) {
+            latency -= LATE_RECYCLE_INTERVAL_SECONDS * 1000L;
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:239]: Why to minus the LATE_RECYCLE_INTERVAL_SECONDS?

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 7]
```diff
@@ -77,7 +78,12 @@ public class CatalogRecycleBin extends MasterDaemon implements Writable {
     private com.google.common.collect.Table<Long, String, RecycleTableInfo> nameToTableInfo;
     private Map<Long, RecyclePartitionInfo> idToPartition;
 
-    private Map<Long, Long> idToRecycleTime;
+    // if enable erase later, the real recycle time is RecycleTime + LATE_RECYCLE_INTERVAL_SECONDS
+    // This is only useful on master, when tablet scheduler repair a tablet that is about to expire
+    // We can make the db/table/partition infomation stay longer until the asynchronized agent task finish.
+    protected static int LATE_RECYCLE_INTERVAL_SECONDS = 60;
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:84]: What's  the purpose of the LATE_RECYCLE_INTERVAL_SECONDS?

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java (hunk 5 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java]: Done

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java (hunk 6 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java]: Changed some line to make the logic clearer.

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 8]
```diff
@@ -217,22 +229,50 @@ public synchronized List<Partition> getPartitions(long tableId) {
...
+     * make sure there are still some time before the subject is erased
+     */
+    public synchronized boolean ensureEraseLater(long id, long currentTimeMs) {
+        // 1. not in idToRecycleTime, maybe already erased, sorry it's too late!
+        if (!idToRecycleTime.containsKey(id)) {
+            return false;
+        }
+        // 2. will expire after quite a long time, don't worry
+        long latency = currentTimeMs - idToRecycleTime.get(id);
+        if (latency < (Config.catalog_trash_expire_second - LATE_RECYCLE_INTERVAL_SECONDS) * 1000L) {
```
[the remark this answers is not in the document]
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:260]: I think as long as the logic is clear, there's no rule that must return the same value. If I have to make sure all branches return true/false, the code will be very unreadable.

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java (hunk 7 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java]: Added some comments to explain why.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 42 / 48 (87.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|14|17|82.35%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|28|31|90.32%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 9]
```diff
@@ -217,22 +233,52 @@ public synchronized List<Partition> getPartitions(long tableId) {
...
+            return false;
+        }
+        // 2. will expire after quite a long time, don't worry
         long latency = currentTimeMs - idToRecycleTime.get(id);
-        return latency > MIN_ERASE_LATENCY && latency > Config.catalog_trash_expire_second * 1000L;
+        if (latency < (Config.catalog_trash_expire_second - LATE_RECYCLE_INTERVAL_SECONDS) * 1000L) {
+            return true;
+        }
+        // 3. still expire after adding recycle time, sorry.
+        if (latency - LATE_RECYCLE_INTERVAL_SECONDS * 1000L > Config.catalog_trash_expire_second * 1000L) {
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:264]: Does the 2 and 3 is equivalent?

Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:264]:
```suggestion
        if (latency > (Config.catalog_trash_expire_second  + LATE_RECYCLE_INTERVAL_SECONDS) * 1000L) {
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 42 / 48 (87.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|14|17|82.35%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|28|31|90.32%|

**Review by chaoyli — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by padmejin** [comment]
@Mer

**Pull request merged**